lib/engines/hcktest/tests.rb

```diff
@@ -305,31 +331,42 @@ def update_summary_results_log
       generate_report
     end
 
-    def update_remote(test_id, test_logs_path, status, testname)
-      delete_old_remote(testname)
-      new_filename = "#{status}: #{testname}"
-      r_name = new_filename + File.extname(test_logs_path)
+    def update_remote(test_id, test_result, test_logs_path, status, testname)
+      test_instance_id = test_result['instanceid']
+      delete_old_remote(testname, test_instance_id)
+
+      r_name = "#{status}_#{test_instance_id}_#{testname}#{File.extname(test_logs_path)}"
       @project.result_uploader.upload_file(test_logs_path, r_name)
 
+      r_name = "Result_#{test_instance_id}_#{testname}.json"
+      File.write("#{@project.workspace_path}/#{r_name}", JSON.pretty_generate(test_result))
+      @project.result_uploader.upload_file("#{@project.workspace_path}/#{r_name}", r_name)
+
       if @tests_extra.dig(test_id, 'dump')
-        r_name = "Minidump: #{testname}.zip"
+        r_name = "Minidump_#{test_instance_id}_#{testname}.zip"
         @project.result_uploader.upload_file(@tests_extra.dig(test_id, 'dump'), r_name)
       end
 
       update_summary_results_log
     end
 
-    def delete_old_remote(test_name)
-      r_name = "Minidump: #{test_name}.zip"
+    def delete_old_remote(test_name, result_instance_id)
+      r_name = "Minidump_#{result_instance_id}_#{test_name}.zip"
+      @project.result_uploader.delete_file(r_name)
+      r_name = "Result_#{result_instance_id}_#{test_name}.json"
       @project.result_uploader.delete_file(r_name)
-      r_name = "Failed: #{test_name}.zip"
+      r_name = "Failed_#{result_instance_id}_#{test_name}.zip"
       @project.result_uploader.delete_file(r_name)
-      r_name = "Passed: #{test_name}.zip"
+      r_name = "Passed_#{result_instance_id}_#{test_name}.zip"
       @project.result_uploader.delete_file(r_name)
     end
 
     def all_tests_finished?
-      status_count('InQueue').zero? && current_test.nil?
+      status_count('InQueue').zero? &&
+        # TestResult.Status does not return Queued
+        # (if a test is scheduled or running it returns Running).
+        results_count('Running').zero? &&
```

Comment:
`status_count` and `results_count` may not be worthwhile to have methods; they are trivial and currently called only in this method. We can also simplify them a bit by using `Array#any?` instead of `Array#count`.